Merge updates from ocm-plus repo

JustinKuli opened the pull request:
Brings in commits up to the current main branch from the other repo.

Major changes include test updates which should make them less flakey in the CI, and the operator-sdk update, which makes more YAML automatically, and updates the controller-runtime and associated libraries.

mprahl committed:
Upgrade to the latest operator-sdk (#54)

* Upgrade to the latest operator-sdk

operator-sdk v1.x now wraps Kubebuilder for its scaffolding and code
generation. This changes the path of `./cmd/manager/main.go` to
`.main.go` and `./pkg/controller` to `./controllers`.

Manifest generation is now handled by kubebuilder. This is handled by
the `generate-operator-yaml` Make target. This means that any RBAC
changes must be done using kubebuilder markers and not manual edits to
the manifests in `./deploy`.

The largest changes stem from updating the underlying Kubernetes Go
modules as some APIs changed or were removed.

It's also important to note that operator-sdk is no longer a Go module
dependency, so some functionality was copied over, mostly unmodified.

Resolves:
https://github.com/open-cluster-management/backlog/issues/15550

Signed-off-by: mprahl <mprahl@users.noreply.github.com>

* Add health and ready probes

Signed-off-by: mprahl <mprahl@users.noreply.github.com>

* Wait for the kubectl command in the e2e tests

This reduces the probability of race conditions occurring.

Signed-off-by: mprahl <mprahl@users.noreply.github.com>

* Validate type assertions in the case3 tests

This prevents panics in the tests when the policy doesn't have a status
field yet.

This also guards against index out of range panics in some of the tests.

Signed-off-by: mprahl <mprahl@users.noreply.github.com>

* Wrap test API updates with eventually

This will help in situations where update returns:
Operation cannot be fulfilled on
policies.policy.open-cluster-management.io
\"default.case1-test-policy\": the object has been modified; please apply
your changes to the latest version and try again

Signed-off-by: mprahl <mprahl@users.noreply.github.com>

mprahl committed:
Fix the main.go path in the Make targets (#55)

This was overlooked as part of the operator-sdk upgrade.

Signed-off-by: mprahl <mprahl@users.noreply.github.com>

mprahl committed:
Fix the tag of the quay.io container image (#56)

The Deployment manifest in `deploy/operator.yaml` was referencing the
incorrect and old tag. This should be using the `latest` tag instead.

Signed-off-by: mprahl <mprahl@users.noreply.github.com>

mprahl committed:
Remove the stale apis package (#57)

This should have been removed as part of the operator-sdk upgrade but it
was accidentally left behind.

Signed-off-by: mprahl <mprahl@users.noreply.github.com>

Commit:
Fix CheckAddonPodFunc for leaseUpdater (#60)

Need to use two separate `CheckAddonPodFunc`s to check
policy-framework and config-policy-controller pod.  The additional
`CheckAddonPodFunc` is temporary until addon framework independently
verifies the config-policy-controller via its lease.
See https://github.com/open-cluster-management/backlog/issues/11508

For: https://github.com/open-cluster-management/backlog/issues/17607

Signed-off-by: Yu Cao <ycao@redhat.com>

Commit:
Use config_checker as the healthz checker (#62)

Use config_checker as the healthz checker so that the pod can be
restarted proactively when hub kubeconfig is changed. This is necessary for backup
and restore scenario.
For issue: https://github.com/open-cluster-management/backlog/issues/17651

Signed-off-by: Yu Cao <ycao@redhat.com>

willkutler committed:
Allow using the legacy leader election method (#64)

* Allow using the legacy leader election method

Signed-off-by: Will Kutler <wkutler@redhat.com>

* Add formatting fixes

Signed-off-by: Will Kutler <wkutler@redhat.com>

JustinKuli committed:
Merge 'midstream' main into 'upstream'

Signed-off-by: Justin Kulikauskas <jkulikau@redhat.com>

mprahl commented on Makefile:
```diff
@@ -219,8 +249,9 @@ e2e-test:
-	go get github.com/onsi/gomega/...@v1.10.1
+	go get github.com/onsi/ginkgo/ginkgo@v1.16.4
+	go get github.com/onsi/gomega/...@v1.13.0
+	go get github.com/open-cluster-management/governance-policy-propagator@v0.0.0-20211012174109-95c3b77cce09
```
Could you please change this to a `-io` reference?

JustinKuli replied:
Go modules are going to be hard for this for now. For example, since the -io propagator declares its module in its go.mod **without** -io, Go gives this error when trying to import it:

```
go: github.com/open-cluster-management-io/governance-policy-propagator@v0.5.0: parsing go.mod:
	module declares its path as: github.com/open-cluster-management/governance-policy-propagator
	        but was required as: github.com/open-cluster-management-io/governance-policy-propagator
```

So, for now, I want to defer that change.

mprahl commented on Makefile:
```diff
@@ -238,7 +269,7 @@ e2e-debug:
 ############################################################
 build-instrumented:
-	go test -covermode=atomic -coverpkg=github.com/open-cluster-management/$(IMG)... -c -tags e2e ./cmd/manager -o build/_output/bin/$(IMG)-instrumented
+	go test -covermode=atomic -coverpkg=github.com/open-cluster-management/$(IMG)... -c -tags e2e ./ -o build/_output/bin/$(IMG)-instrumented
```
```suggestion
	go test -covermode=atomic -coverpkg=github.com/open-cluster-management-io/$(IMG)... -c -tags e2e ./ -o build/_output/bin/$(IMG)-instrumented
```

JustinKuli replied:
I think the package/module name still does not have `-io`. And I think that's correct for now. Eventually I would want to move the module name to be separate from the exact github path, like https://github.com/open-cluster-management-io/registration/blob/main/go.mod#L1. Then when we make our "midstream" fork, we can do `replace` directives in our `go.mod` files to make things point to the correct repos as necessary, without too much mess around URLs.

mprahl replied:
Ah, I see we haven't updated the `go.mod` file in this repo yet. Why hasn't the module name been changed? How would reference this repo vs midstream if the module name is the same?

JustinKuli replied:
Via `replace`: https://golang.org/ref/mod#go-mod-file-replace

With special attention to:
> Regardless of whether a replacement is specified with a local path or module path, if the replacement module has a go.mod file, its module directive must match the module path it replaces.

So the upstream and midstream go.mod should declare the same module name.

Some sort of redirect system would send `go get` commands to one of the repos by default. Then if the "other" version is needed for a project, they would use a `replace` directive to point to that repo.

mprahl commented on PROJECT:
```diff
@@ -0,0 +1,8 @@
+plugins:
+  manifests.sdk.operatorframework.io/v2: {}
+projectName: governance-policy-status-sync
+repo: github.com/open-cluster-management/governance-policy-status-sync
```
```suggestion
repo: github.com/open-cluster-management-io/governance-policy-status-sync
```

mprahl commented on deploy/manager/manager.yaml:
```diff
@@ -0,0 +1,58 @@
+      serviceAccountName: governance-policy-status-sync
+      containers:
+        - name: governance-policy-status-sync
+          image: quay.io/open-cluster-management/governance-policy-status-sync:latest
```
Should this be -io?

JustinKuli replied:
Images from this (`-io`) repo are pushed to quay.io/open-cluster-management, there is not a separate org.

Although, the image tag built here is `edge`, not `latest`, so that should be adjusted.

mprahl commented on deploy/operator.yaml:
```diff
@@ -154,36 +81,103 @@ rules:
+              fieldPath: metadata.name
+        - name: OPERATOR_NAME
+          value: governance-policy-status-sync
+        image: quay.io/open-cluster-management/governance-policy-status-sync:latest
```
Should this be `-io`?

mprahl commented on go.mod:
```diff
@@ -3,20 +3,21 @@ module github.com/open-cluster-management/governance-policy-status-sync
-	github.com/operator-framework/operator-sdk v0.19.4
+	github.com/onsi/ginkgo v1.16.4
+	github.com/onsi/gomega v1.13.0
+	github.com/open-cluster-management/addon-framework v0.0.0-20210621074027-a81f712c10c2
```
This should be `-io`? right?

JustinKuli replied:
Seems like we're importing both? Because the file also has `open-cluster-management.io/addon-framework` ...

I'd like to defer fixing that to another PR (in either repo).

mprahl commented on go.mod:
```diff
@@ -3,20 +3,21 @@ module github.com/open-cluster-management/governance-policy-status-sync
+	github.com/onsi/ginkgo v1.16.4
+	github.com/onsi/gomega v1.13.0
+	github.com/open-cluster-management/addon-framework v0.0.0-20210621074027-a81f712c10c2
+	github.com/open-cluster-management/governance-policy-propagator v0.0.0-20211012174109-95c3b77cce09
```
Could you please use a `-io` reference?

JustinKuli replied:
No 😓 . See other go module discussions.

mprahl commented on go.mod:
```diff
@@ -3,20 +3,21 @@ module github.com/open-cluster-management/governance-policy-status-sync
 replace (
 	github.com/go-logr/zapr => github.com/go-logr/zapr v0.4.0
-	k8s.io/client-go => k8s.io/client-go v0.20.5
+	github.com/open-cluster-management/multicloud-operators-placementrule => github.com/open-cluster-management/multicloud-operators-placementrule v1.2.4-0-20210816-699e5.0.20211012154812-5fac6c25d2f6
```
Is this needed for upstream?

JustinKuli replied:
It pins a specific version of a nested dependency... so I would guess it is still required. I'll test locally.

JustinKuli replied:
Building without it gives a failure.

JustinKuli committed:
Update paths for upstream

This repo pushes an `edge` tag for its image, so the deployment YAML
should also use that (instead of `latest`)

The deployment process should also use CRDs from the propagator repo
in this same organization, for consistency.

Signed-off-by: Justin Kulikauskas <jkulikau@redhat.com>
